File: LeetCode/Java/src/main/java/org/redquark/tutorials/leetcode/LongestCommonPrefix.java
package org.redquark.tutorials.leetcode;

/**
 * @author Anirudh Sharma
 */
public class LongestCommonPrefix {

    /**
     * @param strs - input array of strings
     * @return longest common prefix
     */
    private static String longestCommonPrefix(String[] strs) {
        // Longest common prefix string
        StringBuilder longestCommonPrefix = new StringBuilder();
        // Base condition
        if (strs == null || strs.length == 0) {
            return longestCommonPrefix.toString();
        }
        // Find the minimum length string from the array
        int minimumLength = strs[0].length();
        for (int i = 1; i < strs.length; i++) {
            minimumLength = Math.min(minimumLength, strs[i].length());
        }
        // Loop for the minimum length
        for (int i = 0; i < minimumLength; i++) {
            // Get the current character from first string
            char current = strs[0].charAt(i);
            // Check if this character is found in all other strings or not
            for (String str : strs) {
                if (str.charAt(i) != current) {
                    return longestCommonPrefix.toString();
                }
            }
            longestCommonPrefix.append(current);
        }
        return longestCommonPrefix.toString();
    }

    public static void main(String[] args) {
        System.out.println(longestCommonPrefix(new String[]{"flower", "flow", "flight"}));
        System.out.println(longestCommonPrefix(new String[]{"dog", "racecar", "car"}));
    }
}

File: LeetCode/Kotlin/src/main/kotlin/org/redquark/tutorials/leetcode/LongestCommonPrefix.kt
package org.redquark.tutorials.leetcode

fun longestCommonPrefix(strs: Array<String>): String {
    // Longest common prefix string
    val longestCommonPrefix = StringBuilder()
    // Base condition
    if (strs.isEmpty()) {
        return longestCommonPrefix.toString()
    }
    // Find the minimum length string from the array
    var minimumLength = strs[0].length
    for (i in 1 until strs.size) {
        minimumLength = minimumLength.coerceAtMost(strs[i].length)
    }
    // Loop for the minimum length
    for (i in 0 until minimumLength) {
        // Get the current character from first string
        val current = strs[0][i]
        // Check if this character is found in all other strings or not
        for (str in strs) {
            if (str[i] != current) {
                return longestCommonPrefix.toString()
            }
        }
        longestCommonPrefix.append(current)
    }
    return longestCommonPrefix.toString()
}

fun main() {
    println(longestCommonPrefix(arrayOf("flower", "flow", "flight")))
    println(longestCommonPrefix(arrayOf("dog", "racecar", "car")))
}